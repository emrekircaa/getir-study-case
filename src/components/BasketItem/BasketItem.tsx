import React from "react";
import styled from "styled-components";
import Minus from "../../assets/icons/Minus";
import { StyledButton } from "../Button/style";
import Impluse from "../../assets/icons/Impluse";
import { colors } from "../../theme";

interface BasketItemProps {
  name: string; // name of the counter item
  price: number; // price of the counter item
  itemCount: number; // current item count of the counter item
  increase: () => void;
  decrease: () => void;
}
export const BasketItem: React.FC<BasketItemProps> = ({
  name,
  price,
  itemCount,
  increase,
  decrease,
}) => {
  return (
    <StyledBasketItemContainer data-testid="basket-item-container">
      <StyledProduct>
        <StyledProductName>{name}</StyledProductName>
        <StyledProductPrice>â‚º{price}</StyledProductPrice>
      </StyledProduct>
      <StyledCounter>
        <StyledCountActionButton
          data-testid="decrease-button"
          onClick={() => decrease()}
        >
          <Minus />
        </StyledCountActionButton>
        <StyledCount>{itemCount}</StyledCount>
        <StyledCountActionButton
          data-testid="increase-button"
          onClick={() => increase()}
        >
          <Impluse />
        </StyledCountActionButton>
      </StyledCounter>
    </StyledBasketItemContainer>
  );
};

export const StyledBasketItemContainer = styled.div`
  display: flex;
  width: 100%;
  justify-content: space-between;
  border-bottom: 1.02px solid ${colors.gray50};
  padding-bottom: 18px;
  padding-top: 18px;
  &:first-child {
    padding-top: 0px;
  }
`;
export const StyledProduct = styled.div`
  display: flex;
  flex-direction: column;
  gap: 4px;
`;
export const StyledProductName = styled.div`
  font-size: 14px;
  font-weight: 400;
  line-height: 18px;
  letter-spacing: 0.16px;
  color: ${colors.gray700};
`;

export const StyledProductPrice = styled.div`
  font-weight: 600;
  font-size: 14px;
  line-height: 18px;
  letter-spacing: 0.16px;
  line-break: break-word;
  color: ${colors.primary};
`;

export const StyledCounter = styled.div`
  display: flex;
  width: 74px;
  height: 32.7px;
  justify-content: space-between;
  align-items: center;
  user-select: none;
  margin-right: 7px;
`;

export const StyledCount = styled.div`
  display: flex;
  align-items: center;
  justify-content: center;
  width: 32px;
  height: 32px;
  font-size: 15px;
  font-weight: 700;
  line-height: 20px;
  letter-spacing: 0px;
  background-color: ${colors.primary};
  color: white;
`;

export const StyledCountActionButton = styled(StyledButton)`
  background-color: transparent !important;
  width: auto !important;
`;
