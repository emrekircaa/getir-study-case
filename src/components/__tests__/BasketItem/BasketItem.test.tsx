import { render, fireEvent } from "@testing-library/react";
import { BasketItem } from "../../BasketItem/BasketItem";

describe("BasketItem component", () => {
  it("renders with correct props", () => {
    const { getByTestId, getByText } = render(
      <BasketItem
        name="Sample Item"
        price={10}
        itemCount={3}
        increase={jest.fn()}
        decrease={jest.fn()}
      />
    );

    const basketItemContainer = getByTestId("basket-item-container");
    expect(basketItemContainer).toMatchSnapshot();
  });

  it("calls increase and decrease functions correctly", () => {
    const name = "Sample Item";
    const price = 10;
    const itemCount = 3;
    const increaseMock = jest.fn();
    const decreaseMock = jest.fn();

    const { getByTestId } = render(
      <BasketItem
        name={name}
        price={price}
        itemCount={itemCount}
        increase={increaseMock}
        decrease={decreaseMock}
      />
    );

    const increaseButton = getByTestId("increase-button");
    const decreaseButton = getByTestId("decrease-button");

    fireEvent.click(increaseButton);
    expect(increaseMock).toHaveBeenCalled();

    fireEvent.click(decreaseButton);
    expect(decreaseMock).toHaveBeenCalled();
  });
});
