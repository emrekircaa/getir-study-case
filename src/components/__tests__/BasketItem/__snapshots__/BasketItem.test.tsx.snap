// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BasketItem component renders with correct props 1`] = `
<div
  class="sc-gEvEer kyMQUm"
  data-testid="basket-item-container"
>
  <div
    class="sc-eqUAAy geJcCj"
  >
    <div
      class="sc-fqkvVR chovyL"
    >
      Sample Item
    </div>
    <div
      class="sc-dcJsrY epQLBC"
    >
      â‚º
      10
    </div>
  </div>
  <div
    class="sc-iGgWBj eCASMi"
  >
    <button
      class="sc-aXZVg sc-kAyceB jHQvpS kZetFM"
      data-testid="decrease-button"
    >
      <svg
        fill="none"
        height="3"
        viewBox="0 0 10 3"
        width="10"
        xmlns="http://www.w3.org/2000/svg"
      >
        <rect
          fill="#1EA4CE"
          height="2.04384"
          rx="1"
          width="10"
          y="0.725586"
        />
      </svg>
    </button>
    <div
      class="sc-gsFSXq inSQFr"
    >
      3
    </div>
    <button
      class="sc-aXZVg sc-kAyceB jHQvpS kZetFM"
      data-testid="increase-button"
    >
      <svg
        fill="none"
        height="12"
        width="12"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M6.44 10.407v-4.21h4.12c.243 0 .44-.2.44-.45a.445.445 0 0 0-.44-.45H6.44V1.088A.445.445 0 0 0 6 .638a.445.445 0 0 0-.44.45v4.21H1.44a.445.445 0 0 0-.44.45c0 .248.197.45.44.45h4.12v4.209c0 .248.197.45.44.45s.44-.201.44-.45"
          stroke="#1EA4CE"
        />
      </svg>
    </button>
  </div>
</div>
`;
